Fix signing of a transaction in `compatible shelley transaction signed-transaction` command by not using an incomplete body for signing

carbolymer opened the pull request:
# Changelog

```yaml
- description: |
    Update [cardano-api-10.10.0.0](https://github.com/IntersectMBO/cardano-api/blob/master/cardano-api/CHANGELOG.md#101000)
    Fix signing of a transaction in `compatible shelley transaction signed-transaction` command. Previously two different transaction bodies were used for the resulting transaction and the signature - now it's used the same for both purposes.
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - refactoring    # QoL changes
   - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

This PR contains two commits:
- the actual fix
- the split of Cardano.CLI.Compatible.Transaction into separate modules aligning with cardano-cli design.

Fixes:
- https://github.com/IntersectMBO/cardano-cli/issues/1054

Requires:
- https://github.com/IntersectMBO/cardano-api/pull/750

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

<!--
### Note on CI ###
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges. Please contact IOG node developers to do this
for you.
-->

Jimbo4350 commented on cardano-cli/cardano-cli.cabal:
```diff
@@ -325,7 +325,7 @@ test-suite cardano-cli-test
     base16-bytestring,
     bech32 >=1.1.0,
     bytestring,
-    cardano-api:{cardano-api, gen, internal},
```
:partying_face:

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/Read.hs:
```diff
@@ -519,6 +522,22 @@ data ShelleyBootstrapWitnessSigningKeyData
+  :: ()
+  => ShelleyBasedEra era
+  -> Maybe NetworkId
+  -> A.TxBody era
```
:man_facepalming: It confused me at first but I'm just noticing this type. What do we gain by wrapping the ledger type here? Why not directly use the ledger type?

carbolymer replied:
That we don't have to use `ShelleyLedgerEra` type family and we can hide some calls to `shelleyBasedEraConstraints` behind our lens. A little convenience for some boilerplate.

If we decide to not use this `TxBody` type, we should deprecate the module `Cardano.Api.Ledger.Lens` then to avoid confusion.

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs:
```diff
@@ -279,69 +277,29 @@ runCompatibleTransactionCmd
         )
         sbe
 
-    let certsRefInputs =
```
I assume we don't need reference inputs because QA does not use them?

carbolymer replied:
We do - but removed code was only used to build a body only for creating signatures. But it's problematic: it's a slightly different body than the one we're using in building compatible transaction. So the transaction building logic is duplicated in two places, and we had the bug with incorrect body hash used for the signatures.

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs:
```diff
@@ -279,69 +277,29 @@ runCompatibleTransactionCmd
     let txCerts = mkTxCertificates sbe certsAndMaybeScriptWits
 
-    -- this body is only for witnesses
-    apiTxBody <-
```
:+1:

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/Compatible/Transaction.hs:
```diff
@@ -279,69 +277,29 @@ runCompatibleTransactionCmd
+      firstExceptT CompatiblePParamsConversionError . hoistEither $
+        createCompatibleTx sbe ins allOuts fee protocolUpdates votes txCerts
+
+    let txBody = ledgerTx ^. A.txToTxBodyL sbe
```
We need to directly use the ledger tx body type instead.

carbolymer replied:
https://github.com/IntersectMBO/cardano-cli/pull/1057/files#r1959731654

Jimbo4350 reviewed (commented):
LGTM but I have a few questions

Jimbo4350 reviewed (changes requested):
Fix signing how? Again, break this up in to sensible commits to make it easy for the reviewer to understand what has changed as there are lots of style related changes.

carbolymer committed:
Move mkShelleyBootstrapWitnesses to Cardano.CLI.Read to make it reusable

carbolymer committed:
Compatible tx building: build only one tx body instead of two

This avoids of implementation drift between cardano-api and cardano-cli.
We were building two transaction bodies - one for just signatures and
one is the resulting transaction. Thos bodies over time became different
and resulted in a bug producing invalid transaction signatures.

Now it's only one body used for both purposes.

carbolymer committed:
Add reference test for compatible transaction building

carbolymer committed:
Update cardano-api-10.10.0.0
